### ISSUE TITLE
Is there any function equivalent to tibbletime::as_period()? 

### ISSUE THREAD

**Issue by GitHunter0** [business-science/timetk#64]
Is there any function in `timetk ` equivalent to `tibbletime::as_period()`? 
It is a highly useful functionality that I cannot find elsewhere outside the deprecated `tibbletime `package. 
Do you plan to incorporate it in `timetk`?
Thanks a lot.

**Comment by mdancho84** [comment]
Timetk has summarise_by_time() which is what I normally use. But if that doesn't work for you please provide a minimal example and I can determine if it makes sense to incorporate.

**Comment by GitHunter0** [comment]
@mdancho84 , thank you Matt. `tibbletime::as_period()` does not summarize dataframes, it filters rows by date criterium. I believe it is an essential functionality which complements other `timetk `functions. Here some examples:

```
library(tidyverse)
library(tibbletime)
  
data <- 
  FANG %>% 
    arrange(symbol, date) %>% 
    group_by(symbol) %>% 
    tibbletime::as_tbl_time(index=date)

# Keep observations of the first available day of each month
data %>% tibbletime::as_period("1 month", side="start") %>% View()
# Keep observations of the last available day of the month for every 2 months
data %>% tibbletime::as_period("2 month", side="end") %>% View()
# Keep just the last observation for each year
data %>% tibbletime::as_period("1 year", side="end") %>% View()
# Keep the first observation of each 20 day-interval 
data %>% tibbletime::as_period("20 days", side="end") %>% View() 
```

**Comment by mdancho84** [comment]
You are correct. I've added:
-  `condense_period()` function that enables reducing the periodicity from higher frequency (e.g. daily) to lower frequency (e.g. monthly). 
- `filter_period()`: Applies filtering expressions within time-based periods (windows). 
- `slice_period()`: Applies slices within time-based periods (windows).

Here's a comparison between the `tibbletime::as_period()` and `timetk::condense_period()`:

``` r
library(tidyverse)
library(timetk)
library(tibbletime)

data <- FANG %>% 
    arrange(symbol, date) %>% 
    group_by(symbol) %>% 
    tibbletime::as_tbl_time(index=date)

# Keep observations of the first available day of each month
data %>% tibbletime::as_period("1 month", side="start")
#> # A time tibble: 192 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-01-02  256.  258.  253.  257. 3271000     257.
#>  2 AMZN   2013-02-01  269.  269.  263.  265  6115000     265 
#>  3 AMZN   2013-03-01  263.  267.  261.  266. 2957000     266.
#>  4 AMZN   2013-04-01  267.  267.  261.  262. 2524800     262.
#>  5 AMZN   2013-05-01  254.  254.  246.  248. 4391400     248.
#>  6 AMZN   2013-06-03  269.  269.  263.  267. 2543900     267.
#>  7 AMZN   2013-07-01  279   283.  277.  282. 2888200     282.
#>  8 AMZN   2013-08-01  303.  306.  299.  306. 2971700     306.
#>  9 AMZN   2013-09-03  285.  291.  284.  289. 2897600     289.
#> 10 AMZN   2013-10-01  314.  321   314.  321. 2396400     321.
#> # … with 182 more rows
data %>% condense_period(.by = "1 month", .side = "start")
#> .date_var is missing. Using: date
#> # A time tibble: 192 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-01-02  256.  258.  253.  257. 3271000     257.
#>  2 AMZN   2013-02-01  269.  269.  263.  265  6115000     265 
#>  3 AMZN   2013-03-01  263.  267.  261.  266. 2957000     266.
#>  4 AMZN   2013-04-01  267.  267.  261.  262. 2524800     262.
#>  5 AMZN   2013-05-01  254.  254.  246.  248. 4391400     248.
#>  6 AMZN   2013-06-03  269.  269.  263.  267. 2543900     267.
#>  7 AMZN   2013-07-01  279   283.  277.  282. 2888200     282.
#>  8 AMZN   2013-08-01  303.  306.  299.  306. 2971700     306.
#>  9 AMZN   2013-09-03  285.  291.  284.  289. 2897600     289.
#> 10 AMZN   2013-10-01  314.  321   314.  321. 2396400     321.
#> # … with 182 more rows

# Keep observations of the last available day of the month for every 2 months
data %>% tibbletime::as_period("2 month", side="end")
#> # A time tibble: 96 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-02-28  262.  267   261.  264. 2667200     264.
#>  2 AMZN   2013-04-30  249.  255.  249.  254. 4248800     254.
#>  3 AMZN   2013-06-28  276.  280.  276.  278. 3193300     278.
#>  4 AMZN   2013-08-30  285.  285.  280   281. 1534700     281.
#>  5 AMZN   2013-10-31  362.  366   359   364. 2466800     364.
#>  6 AMZN   2013-12-31  395.  399.  394.  399. 1996500     399.
#>  7 AMZN   2014-02-28  361.  366.  357.  362. 3882000     362.
#>  8 AMZN   2014-04-30  298.  305.  298.  304. 4088600     304.
#>  9 AMZN   2014-06-30  325.  327.  322.  325. 2671300     325.
#> 10 AMZN   2014-08-29  342.  342.  338.  339. 1939500     339.
#> # … with 86 more rows
data %>% condense_period(.by = "2 month", .side = "end")
#> .date_var is missing. Using: date
#> # A time tibble: 96 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-02-28  262.  267   261.  264. 2667200     264.
#>  2 AMZN   2013-04-30  249.  255.  249.  254. 4248800     254.
#>  3 AMZN   2013-06-28  276.  280.  276.  278. 3193300     278.
#>  4 AMZN   2013-08-30  285.  285.  280   281. 1534700     281.
#>  5 AMZN   2013-10-31  362.  366   359   364. 2466800     364.
#>  6 AMZN   2013-12-31  395.  399.  394.  399. 1996500     399.
#>  7 AMZN   2014-02-28  361.  366.  357.  362. 3882000     362.
#>  8 AMZN   2014-04-30  298.  305.  298.  304. 4088600     304.
#>  9 AMZN   2014-06-30  325.  327.  322.  325. 2671300     325.
#> 10 AMZN   2014-08-29  342.  342.  338.  339. 1939500     339.
#> # … with 86 more rows

# Keep just the last observation for each year
data %>% tibbletime::as_period("1 year", side="end") 
#> # A time tibble: 16 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date         open   high    low  close   volume adjusted
#>    <chr>  <date>      <dbl>  <dbl>  <dbl>  <dbl>    <dbl>    <dbl>
#>  1 AMZN   2013-12-31  395.   399.   394.   399.   1996500    399. 
#>  2 AMZN   2014-12-31  312.   313.   310.   310.   2048000    310. 
#>  3 AMZN   2015-12-31  686.   688.   676.   676.   3718200    676. 
#>  4 AMZN   2016-12-30  766.   767.   748.   750.   4125300    750. 
#>  5 FB     2013-12-31   54.1   54.9   53.9   54.7 43076200     54.7
#>  6 FB     2014-12-31   79.5   79.8   77.9   78.0 19935400     78.0
#>  7 FB     2015-12-31  106    106.   105.   105.  18298700    105. 
#>  8 FB     2016-12-30  117.   117.   115.   115.  18600100    115. 
#>  9 GOOG   2013-12-31 1112.  1121.  1106.  1121.   2725900    560. 
#> 10 GOOG   2014-12-31  531.   533.   526.   526.   1368200    526. 
#> 11 GOOG   2015-12-31  770.   770.   758.   759.   1489600    759. 
#> 12 GOOG   2016-12-30  783.   783.   770.   772.   1760200    772. 
#> 13 NFLX   2013-12-31  366.   369.   364.   368.  10516800     52.6
#> 14 NFLX   2014-12-31  343.   346.   342.   342.   8627500     48.8
#> 15 NFLX   2015-12-31  116.   117.   114.   114.   9245000    114. 
#> 16 NFLX   2016-12-30  126.   127.   124.   124.   4426500    124.
data %>% condense_period(.by = "1 year", .side = "end") 
#> .date_var is missing. Using: date
#> # A time tibble: 16 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date         open   high    low  close   volume adjusted
#>    <chr>  <date>      <dbl>  <dbl>  <dbl>  <dbl>    <dbl>    <dbl>
#>  1 AMZN   2013-12-31  395.   399.   394.   399.   1996500    399. 
#>  2 AMZN   2014-12-31  312.   313.   310.   310.   2048000    310. 
#>  3 AMZN   2015-12-31  686.   688.   676.   676.   3718200    676. 
#>  4 AMZN   2016-12-30  766.   767.   748.   750.   4125300    750. 
#>  5 FB     2013-12-31   54.1   54.9   53.9   54.7 43076200     54.7
#>  6 FB     2014-12-31   79.5   79.8   77.9   78.0 19935400     78.0
#>  7 FB     2015-12-31  106    106.   105.   105.  18298700    105. 
#>  8 FB     2016-12-30  117.   117.   115.   115.  18600100    115. 
#>  9 GOOG   2013-12-31 1112.  1121.  1106.  1121.   2725900    560. 
#> 10 GOOG   2014-12-31  531.   533.   526.   526.   1368200    526. 
#> 11 GOOG   2015-12-31  770.   770.   758.   759.   1489600    759. 
#> 12 GOOG   2016-12-30  783.   783.   770.   772.   1760200    772. 
#> 13 NFLX   2013-12-31  366.   369.   364.   368.  10516800     52.6
#> 14 NFLX   2014-12-31  343.   346.   342.   342.   8627500     48.8
#> 15 NFLX   2015-12-31  116.   117.   114.   114.   9245000    114. 
#> 16 NFLX   2016-12-30  126.   127.   124.   124.   4426500    124.

# Keep the first observation of each 20 day-interval 
data %>% tibbletime::as_period("20 days", side="start") 
#> # A time tibble: 292 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-01-02  256.  258.  253.  257. 3271000     257.
#>  2 AMZN   2013-01-22  272.  272.  269.  270. 2137700     270.
#>  3 AMZN   2013-02-11  263.  263.  257.  257. 3403700     257.
#>  4 AMZN   2013-03-04  265.  273.  264.  273. 3453000     273.
#>  5 AMZN   2013-03-22  255.  258.  252.  258. 3124600     258.
#>  6 AMZN   2013-04-11  265.  271.  265.  270. 3090500     270.
#>  7 AMZN   2013-05-01  254.  254.  246.  248. 4391400     248.
#>  8 AMZN   2013-05-21  269.  271.  268.  269. 1741200     269.
#>  9 AMZN   2013-06-10  277.  282.  275.  281. 3086500     281.
#> 10 AMZN   2013-07-01  279   283.  277.  282. 2888200     282.
#> # … with 282 more rows
data %>% condense_period(.by = "20 days", .side = "start")
#> .date_var is missing. Using: date
#> # A time tibble: 384 x 8
#> # Index:  date
#> # Groups: symbol [4]
#>    symbol date        open  high   low close  volume adjusted
#>    <chr>  <date>     <dbl> <dbl> <dbl> <dbl>   <dbl>    <dbl>
#>  1 AMZN   2013-01-02  256.  258.  253.  257. 3271000     257.
#>  2 AMZN   2013-01-22  272.  272.  269.  270. 2137700     270.
#>  3 AMZN   2013-02-01  269.  269.  263.  265  6115000     265 
#>  4 AMZN   2013-02-21  265.  269.  263.  266. 3637400     266.
#>  5 AMZN   2013-03-01  263.  267.  261.  266. 2957000     266.
#>  6 AMZN   2013-03-21  256.  257   253.  253. 2806400     253.
#>  7 AMZN   2013-04-01  267.  267.  261.  262. 2524800     262.
#>  8 AMZN   2013-04-22  259.  265.  258.  264. 2119100     264.
#>  9 AMZN   2013-05-01  254.  254.  246.  248. 4391400     248.
#> 10 AMZN   2013-05-21  269.  271.  268.  269. 1741200     269.
#> # … with 374 more rows
```

<sup>Created on 2020-11-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mdancho84** [comment]
I'm going to close this. Let me know if you see anything else.

**Comment by GitHunter0** [comment]
@mdancho84 Hey Matt, you never disappoints! 
You are doing such an unbelievable job in all your packages. `timetk `, for example, now that looked more into it, it seems the best time series and longitudinal / panel data package I have ever seen.

As regarding the present issue, just a quick follow up. 
Looking closely, we can see that the results in the last example differ from the tibbletime's, it seems that `timetk::condense_period("20 days")` is incorrectly grouping by month.

Lastly, since pad_by_time() already has a bult-in function to detect/guess the periodicity of the data, could you build a separate function for that? It could work by creating an additional column indicating the guessed periodicity for each serie. It is a very useful functionality in some situations (so much that I built my own function for that).

Thank you

**Comment by mdancho84** [comment]
Thanks, yes - `timetk` handles everything. It's a fairly complete system for handling the bulk of time series data wrangling, visualization, and preprocessing needs. 

The `condense_period()` function leverages `floor_date()` to find the groups. If "20 days" causes an incorrect calculation, then we need to file an issue with `lubridate`, since that's the underlying function. I'd like to keep the `lubridate` functionality since it works for most cases and really should be updated on their end if there's an issue.

**Comment by mdancho84** [comment]
A potential solution is to implement `tibbletime::partition_index()`. I don't have the bandwidth at the moment, but this is an idea.

**Comment by GitHunter0** [comment]
Thank you @mdancho84 . Following your suggestion, I opened an issue with `lubridate` here:
https://github.com/tidyverse/lubridate/issues/932

**Comment by mdancho84** [comment]
Ok, thanks for this - This is the right way to go. I know several users have complained about odd periods when `summarize_by_time()` (also uses `floor_date()`) gives incorrect periods. This should fix all of them.

**Comment by mdancho84** [comment]
Just a foll-up

**Comment by mdancho84** [comment]
Just a follow-up on this - I'm adding a link to `clock` a new package that improves some of the `lubridate` functionality such as `floor_date()`. https://github.com/r-lib/clock/

**Comment by GitHunter0** [comment]
Thanks Matt, I could not test it due to an installation error, so I opened an issue and will get back to it as soon as that is solved

**Comment by GitHunter0** [comment]
Now I managed to test `clock` functions `date_ceiling()` and `date_floor()`, they seem to be very useful for other purposes. However they does not solve our issue here (at least I could not). Let's hope `lubridate` addresses it in the near future.

Anyway, `clock` seems a promising package, thanks for the suggestion, @mdancho84

**Comment by mdancho84** [comment]
@DavisVaughan I wanted to get your thoughts on a challenge with `floor_date()` and `ceiling_date()`. The default incorrectly works from an origin starting at the month level. With your work on `clock`, which looks very promising, I wanted to see if this incorrect `lubridate` behavior is addressed. Ref: tidyverse/lubridate#932

``` r
library(lubridate)
#> 
#> Attaching package: 'lubridate'
#> The following objects are masked from 'package:base':
#> 
#>     date, intersect, setdiff, union

dates <- seq.Date(as.Date("2011-01-01"), by = "day", length.out = 365*2)

table(floor_date(dates, unit = "20 days"))
#> 
#> 2011-01-01 2011-01-21 2011-02-01 2011-02-21 2011-03-01 2011-03-21 2011-04-01 
#>         20         11         20          8         20         11         20 
#> 2011-04-21 2011-05-01 2011-05-21 2011-06-01 2011-06-21 2011-07-01 2011-07-21 
#>         10         20         11         20         10         20         11 
#> 2011-08-01 2011-08-21 2011-09-01 2011-09-21 2011-10-01 2011-10-21 2011-11-01 
#>         20         11         20         10         20         11         20 
#> 2011-11-21 2011-12-01 2011-12-21 2012-01-01 2012-01-21 2012-02-01 2012-02-21 
#>         10         20         11         20         11         20          9 
#> 2012-03-01 2012-03-21 2012-04-01 2012-04-21 2012-05-01 2012-05-21 2012-06-01 
#>         20         11         20         10         20         11         20 
#> 2012-06-21 2012-07-01 2012-07-21 2012-08-01 2012-08-21 2012-09-01 2012-09-21 
#>         10         20         11         20         11         20         10 
#> 2012-10-01 2012-10-21 2012-11-01 2012-11-21 2012-12-01 2012-12-21 
#>         20         11         20         10         20         10
```

<sup>Created on 2021-02-19 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by mdancho84** [comment]
@GitHunter0 Looks to me like `clock` solves it. 

``` r
library(lubridate)
#> 
#> Attaching package: 'lubridate'
#> The following objects are masked from 'package:base':
#> 
#>     date, intersect, setdiff, union
library(clock)

dates <- seq.Date(as.Date("2011-01-01"), by = "day", length.out = 365*2)

table(floor_date(dates, unit = "20 days"))
#> 
#> 2011-01-01 2011-01-21 2011-02-01 2011-02-21 2011-03-01 2011-03-21 2011-04-01 
#>         20         11         20          8         20         11         20 
#> 2011-04-21 2011-05-01 2011-05-21 2011-06-01 2011-06-21 2011-07-01 2011-07-21 
#>         10         20         11         20         10         20         11 
#> 2011-08-01 2011-08-21 2011-09-01 2011-09-21 2011-10-01 2011-10-21 2011-11-01 
#>         20         11         20         10         20         11         20 
#> 2011-11-21 2011-12-01 2011-12-21 2012-01-01 2012-01-21 2012-02-01 2012-02-21 
#>         10         20         11         20         11         20          9 
#> 2012-03-01 2012-03-21 2012-04-01 2012-04-21 2012-05-01 2012-05-21 2012-06-01 
#>         20         11         20         10         20         11         20 
#> 2012-06-21 2012-07-01 2012-07-21 2012-08-01 2012-08-21 2012-09-01 2012-09-21 
#>         10         20         11         20         11         20         10 
#> 2012-10-01 2012-10-21 2012-11-01 2012-11-21 2012-12-01 2012-12-21 
#>         20         11         20         10         20         10

table(date_floor(dates, precision = "day", n = 20))
#> 
#> 2010-12-17 2011-01-06 2011-01-26 2011-02-15 2011-03-07 2011-03-27 2011-04-16 
#>          5         20         20         20         20         20         20 
#> 2011-05-06 2011-05-26 2011-06-15 2011-07-05 2011-07-25 2011-08-14 2011-09-03 
#>         20         20         20         20         20         20         20 
#> 2011-09-23 2011-10-13 2011-11-02 2011-11-22 2011-12-12 2012-01-01 2012-01-21 
#>         20         20         20         20         20         20         20 
#> 2012-02-10 2012-03-01 2012-03-21 2012-04-10 2012-04-30 2012-05-20 2012-06-09 
#>         20         20         20         20         20         20         20 
#> 2012-06-29 2012-07-19 2012-08-08 2012-08-28 2012-09-17 2012-10-07 2012-10-27 
#>         20         20         20         20         20         20         20 
#> 2012-11-16 2012-12-06 2012-12-26 
#>         20         20          5
```

<sup>Created on 2021-02-19 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by GitHunter0** [comment]
> @GitHunter0 Looks to me like `clock` solves it.

@mdancho84 , it is because of your particular example. See this counter-example:

```
dates <- seq.Date(as.Date("2014-07-03"), by = "day", length.out = 28) ; dates
#>  [1] "2014-07-03" "2014-07-04" "2014-07-05" "2014-07-06" "2014-07-07"
#>  [6] "2014-07-08" "2014-07-09" "2014-07-10" "2014-07-11" "2014-07-12"
#> [11] "2014-07-13" "2014-07-14" "2014-07-15" "2014-07-16" "2014-07-17"
#> [16] "2014-07-18" "2014-07-19" "2014-07-20" "2014-07-21" "2014-07-22"
#> [21] "2014-07-23" "2014-07-24" "2014-07-25" "2014-07-26" "2014-07-27"
#> [26] "2014-07-28" "2014-07-29" "2014-07-30"

clock::date_floor(dates, precision = "day", n = 20)
#>  [1] "2014-06-19" "2014-06-19" "2014-06-19" "2014-06-19" "2014-06-19"
#>  [6] "2014-06-19" "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09"
#> [11] "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09"
#> [16] "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09"
#> [21] "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09" "2014-07-09"
#> [26] "2014-07-09" "2014-07-29" "2014-07-29"
```

The result, instead of starting in `2014-07-03`, starts at `2014-06-19`

**Comment by DavisVaughan** [comment]
The tl;dr is that clock is able to solve this problem through `date_floor()`, possibly with a shifted `origin` argument if required.

The behavior in lubridate is not a bug. It is working as expected. You just have to reframe your way of thinking about the problem. To understand what lubridate does, it is easier to explain how clock works and then compare.

In clock, summarizing time has been broken into 3 families of functionality:

- Grouping: `date_group()`
- Rounding: `date_floor()`, `date_round()`, `date_ceiling()`
- Shifting: `date_shift()`

These all summarize time in slightly different ways.

Grouping lets you group _within larger units of time_. So when you group by `"day"` on a Date (or year-month-day clock type), really you are grouping by the "day of the current year-month". This means that grouping cannot cross month boundaries, and the count of "20 days" restarts at the beginning of each month.

Rounding uses mathematical definitions of floor and ceiling. For Date objects, it operates on the underlying integer count of days. Flooring will summarize rolling groups of 20 days starting from the implied origin of 1970-01-01, crossing month and year boundaries. So you'd get groups of `[1970-01-01, 1970-01-21), [1970-01-21, 1970-02-10)`, and so on. If you need to change the origin to something other than 1970-01-01, you can use the `origin` argument.

```r
library(clock)

x <- as.Date("1970-01-01") + 0:5
x
#> [1] "1970-01-01" "1970-01-02" "1970-01-03" "1970-01-04" "1970-01-05"
#> [6] "1970-01-06"

unclass(x)
#> [1] 0 1 2 3 4 5

date_floor(x, "day", n = 2)
#> [1] "1970-01-01" "1970-01-01" "1970-01-03" "1970-01-03" "1970-01-05"
#> [6] "1970-01-05"

date_floor(x, "day", n = 2, origin = as.Date("1970-01-02"))
#> [1] "1969-12-31" "1970-01-02" "1970-01-02" "1970-01-04" "1970-01-04"
#> [6] "1970-01-06"
```

Shifting allows you to target a specific weekday, and shift your dates to the next or previous instance of that weekday. This is the easiest way to summarize by week if you have a specific weekday in mind that corresponds to the "first day of the week".

```r
library(clock)

x <- as.Date("1970-01-01") + 0:5

as_weekday(x)
#> <weekday[6]>
#> [1] Thu Fri Sat Sun Mon Tue

sunday <- weekday(clock_weekdays$sunday)
monday <- weekday(clock_weekdays$monday)

# shift to the previous sunday
date_shift(x, target = sunday, which = "previous")
#> [1] "1969-12-28" "1969-12-28" "1969-12-28" "1970-01-04" "1970-01-04"
#> [6] "1970-01-04"

# shift to the previous monday
date_shift(x, target = monday, which = "previous")
#> [1] "1969-12-29" "1969-12-29" "1969-12-29" "1969-12-29" "1970-01-05"
#> [6] "1970-01-05"

# shift to the previous monday
# if we were already on a monday, shift it anyways
date_shift(x, target = monday, which = "previous", boundary = "advance")
#> [1] "1969-12-29" "1969-12-29" "1969-12-29" "1969-12-29" "1969-12-29"
#> [6] "1970-01-05"
```

lubridate `floor_date()` uses grouping behavior for everything except "weeks", which uses shifting behavior. I think that the clock functions are a bit more consistent.

A lot more information at:
https://r-lib.github.io/clock/articles/clock.html#grouping-rounding-and-shifting-1

**Comment by DavisVaughan** [comment]
So @GitHunter0 for that example you'd need to use the origin if you really wanted the start of your series to be where the 20 day count starts from. In general it is probably best not to care about the origin too much. I find that having a fixed origin of 1970-01-01 that is _always_ used is more consistent than having an origin that changes based on your input data.

```r
library(clock)

origin <- as.Date("2014-07-03")
dates <- seq.Date(origin, by = "day", length.out = 28)

date_floor(dates, precision = "day", n = 20, origin = origin)
#>  [1] "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03"
#>  [6] "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03"
#> [11] "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03"
#> [16] "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03" "2014-07-03"
#> [21] "2014-07-23" "2014-07-23" "2014-07-23" "2014-07-23" "2014-07-23"
#> [26] "2014-07-23" "2014-07-23" "2014-07-23"
```

**Comment by GitHunter0** [comment]
Thank you for taking the time to respond, @DavisVaughan. 

The origin option solves part of the problem. The complete issue we discussed in tidyverse/lubridate#932 is that I don't want any output date that is not present in the input dates, it should approximate to the closest date existing in the original input of dates. 

Here is an example:
```
dates <- as.Date(c("2014-07-03", "2014-07-20", "2014-07-24"))
clock::date_floor(dates, precision = "day", n = 20, origin = dates[1])
#> [1] "2014-07-03" "2014-07-03" "2014-07-23"
```

In this case, the output `2014-07-23` does not exist in the input dates, so it should be approximated to the closest date, which is `2014-07-24`.

**Comment by DavisVaughan** [comment]
I don't think that `date_floor()` and friends will take on the job of this approximation behavior, but something like this may help you:

``` r
library(clock)
library(dplyr)

x <- as.Date(c(
  '2000-03-04', 
  '2000-03-22',
  '2000-03-25',
  '2000-04-12',
  '2000-04-15',
  '2000-04-29'
)) 

df <- tibble(x = x) %>%
  mutate(
    lower = date_floor(x, "day", n = 20, origin = x[1]),
    upper = date_ceiling(x + 1, "day", n = 20, origin = x[1])
  )

df %>%
  group_by(lower, upper) %>%
  mutate(x_min_in_range = min(x))
#> # A tibble: 6 x 4
#> # Groups:   lower, upper [3]
#>   x          lower      upper      x_min_in_range
#>   <date>     <date>     <date>     <date>        
#> 1 2000-03-04 2000-03-04 2000-03-24 2000-03-04    
#> 2 2000-03-22 2000-03-04 2000-03-24 2000-03-04    
#> 3 2000-03-25 2000-03-24 2000-04-13 2000-03-25    
#> 4 2000-04-12 2000-03-24 2000-04-13 2000-03-25    
#> 5 2000-04-15 2000-04-13 2000-05-03 2000-04-15    
#> 6 2000-04-29 2000-04-13 2000-05-03 2000-04-15
```

<sup>Created on 2021-02-19 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by GitHunter0** [comment]
Great man, nice workaround @DavisVaughan

**Comment by mdancho84** [comment]
Thanks a lot for your guidance @DavisVaughan

**Comment by mdancho84** [comment]
Closing this.
